x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java

```diff
@@ -292,30 +297,41 @@ private void executeCompletionListeners() {
         }
         // we don't need to restore the response headers, they should be included in the current
         // context since we are called by the search action listener.
-        AsyncSearchResponse finalResponse = getResponse();
-        for (Consumer<AsyncSearchResponse> listener : completionListeners.values()) {
-            listener.accept(finalResponse);
-        }
-        completionListeners.clear();
+        getResponse(completionListeners);
+        //TODO is clearing the map necessary? we will only execute the listeners once anyways
+        //completionListeners.clear();
     }
 
     /**
      * Returns the current {@link AsyncSearchResponse}.
      */
-    private AsyncSearchResponse getResponse() {
-        assert searchResponse.get() != null;
-        checkCancellation();
-        return searchResponse.get().toAsyncSearchResponse(this, expirationTimeMillis);
+    private void getResponse(ActionListener<AsyncSearchResponse> listener) {
+        getResponse(listener, false);
     }
 
     /**
      * Returns the current {@link AsyncSearchResponse} and restores the response headers
      * in the local thread context.
      */
-    private AsyncSearchResponse getResponseWithHeaders() {
+    private void getResponseWithHeaders(ActionListener<AsyncSearchResponse> listener) {
+        getResponse(listener, true);
+    }
+
+    private void getResponse(ActionListener<AsyncSearchResponse> listener, boolean restoreResponseHeaders) {
         assert searchResponse.get() != null;
         checkCancellation();
-        return searchResponse.get().toAsyncSearchResponseWithHeaders(this, expirationTimeMillis);
+        AsyncSearchResponse asyncSearchResponse;
+        try {
+            asyncSearchResponse = searchResponse.get().toAsyncSearchResponse(this, expirationTimeMillis, restoreResponseHeaders);
+        } catch(Exception e) {
+            listener.onFailure(e);
+            return;
+        }
+        listener.onResponse(asyncSearchResponse);
+    }
+
+    AsyncSearchResponse buildErrorResponse(SearchResponse searchResponse, Exception exception) {
```

Comment:
The other solution would be to build the error response when catching the exception in `getResponse` ? This way we don't need to differentiate between a failure during partial reduce and a fatal failure. They both return an async search response that contains a failure ?

Reply:
That is along the lines of what I had in the first iteration, and I believe it defeats the purpose of holding ActionListeners instead of Consumers. We have three places that require different behaviour currently when we get reduction failures:

1) submit when the timeout expires, we treat is a fatal failure and cancel the search task
2) submit when the response completes, we already returned, we store async search response
3) get: we treat it as a transient failure

I think we could maintain the behaviour listed above even if we did not hold action listeners? Only when getting a response the consumer should check if it holds a failure and act accordingly?

I am not sure though if you meant on changing also some of this behaviour, especially because I don't completely follow you statement " This way we don't need to differentiate between a failure during partial reduce and a fatal failure.".

I was actually wondering if we should be more clear with the user, maybe wrapping the exception, and let them know when the exception comes from search and when it comes from async search. I can see how using suppressed exceptions gives you all that happened but it's hard to decipher and debug.

Reply:
What I meant is that the context of the failure is important so we should leave it to the consumer ? In general, and sorry for the back and forth on this, I think we should always return an `AsyncSearchResponse` even in the case of a failure.
The failure can be transient, in such case `is_running` should be `true` (the search action is still running). 
I also don't think we should cancel the search if we have a failure when reducing a partial search response. We should return the response with the transient failure and the metadata associated with the current search (number of shards, id, ...). That would simplify the handling of failures and would be consistent with transient failures in get ?

Reply:
Ok I need to give this a try, especially as the first iteration did not have yet the described different behaviour in submit and get. I think that the main argument for this is to simplify things, adding action listeners adds complexity and having to call buildErrorResponse from submit is weird and should be removed if possible, which is kind of why I initially went down that route :)

Reply:
having gone back to the consumer approach, I see the "context" argument better. Returning async search response is more convenient as it holds the needed info to tell what is happening, while an exception alone does not say much besides that an error has happened.